examples/src/main/java/io/grpc/examples/multiplex/SharingClient.java

```diff
@@ -0,0 +1,194 @@
+        logger.warning("Echo Failed: {0}" + Status.fromThrowable(t));
+        finishLatch.countDown();
+      }
+
+      @Override
+      public void onCompleted() {
+        logger.info("Finished RecordRoute");
+        finishLatch.countDown();
+      }
+    };
+
+    StreamObserver<EchoRequest> requestObserver =
+        echoStub.bidirectionalStreamingEcho(responseObserver);
+    try {
+      for (String curValue : valuesToSend) {
+        EchoRequest req = EchoRequest.newBuilder().setMessage(curValue).build();
+        requestObserver.onNext(req);
+        // Sleep for a bit before sending the next one.
+        Thread.sleep(random.nextInt(1000) + 500);
+        if (finishLatch.getCount() == 0) {
+          // RPC completed or errored before we finished sending.
+          // Sending further requests won't error, but they will just be thrown away.
+          return finishLatch;
+        }
+      }
+    } catch (RuntimeException e) {
+      // Cancel RPC
+      requestObserver.onError(e);
+      throw e;
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      requestObserver.onError(e);
+    }
+    // Mark the end of requests
+    requestObserver.onCompleted();
+
+    return finishLatch;
+  }
+
+  /**
+   * Greet server. If provided, the first element of {@code args} is the name to use in the
+   * greeting. The second argument is the target server.
+   */
+  public static void main(String[] args) throws Exception {
+    String user = "world";
+    // Access a service running on the local machine on port 50051
+    String target = "localhost:50051";
+    // Allow passing in the user and target strings as command line arguments
+    if (args.length > 0) {
+      if ("--help".equals(args[0])) {
+        System.err.println("Usage: [name [target]]");
+        System.err.println("");
+        System.err.println("  name    The name you wish to be greeted by. Defaults to " + user);
+        System.err.println("  target  The server to connect to. Defaults to " + target);
+        System.exit(1);
+      }
+      user = args[0];
+    }
+    if (args.length > 1) {
+      target = args[1];
+    }
+
+    // Create a communication channel to the server, known as a Channel. Channels are thread-safe
+    // and reusable. It is common to create channels at the beginning of your application and reuse
+    // them until the application shuts down.
+    //
+    // For the example we use plaintext insecure credentials to avoid needing TLS certificates. To
+    // use TLS, use TlsChannelCredentials instead.
+    ManagedChannel channel = Grpc.newChannelBuilder(target, InsecureChannelCredentials.create())
+        .build();
+    List<String> echoInput = ImmutableList.of("some", "thing", "wicked", "this", "way", "comes");
+    List<String> echoOutput = new ArrayList<>();
+    try {
+      SharingClient client = new SharingClient(channel);
+
+      CountDownLatch finishLatch = client.initiateEchos(echoInput, echoOutput);
```

Comment:
You might consider using the future stub instead. It makes waiting much more natural.